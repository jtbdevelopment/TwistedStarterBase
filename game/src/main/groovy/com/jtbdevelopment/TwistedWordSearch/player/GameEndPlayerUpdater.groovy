package com.jtbdevelopment.TwistedWordSearch.player

import com.jtbdevelopment.TwistedWordSearch.state.TWSGame
import com.jtbdevelopment.games.dao.AbstractPlayerRepository
import com.jtbdevelopment.games.players.Player
import com.jtbdevelopment.games.publish.GameListener
import com.jtbdevelopment.games.publish.PlayerPublisher
import com.jtbdevelopment.games.state.GamePhase
import groovy.transform.CompileStatic
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

/**
 * Date: 10/5/16
 * Time: 6:39 PM
 */
@Component
@CompileStatic
class GameEndPlayerUpdater implements GameListener<TWSGame> {
    @Autowired
    PlayerPublisher playerPublisher
    @Autowired
    AbstractPlayerRepository playerRepository

    void gameChanged(final TWSGame game, final Player initiatingPlayer, final boolean initiatingServer) {
        if (game.gamePhase == GamePhase.RoundOver && initiatingServer) {

            def playerCount = game.players.size()
            game.players.each {
                Player gamePlayer ->
                    Player p = playerRepository.findOne(gamePlayer.id)
                    TWSPlayerAttributes twsPlayerAttributes = (TWSPlayerAttributes) p.gameSpecificPlayerAttributes
                    if (!twsPlayerAttributes.gamesPlayedByPlayerCount.containsKey(playerCount)) {
                        twsPlayerAttributes.gamesPlayedByPlayerCount[playerCount] = 0
                        twsPlayerAttributes.gamesWonByPlayerCount[playerCount] = 0
                        twsPlayerAttributes.maxScoreByPlayerCount[playerCount] = 0
                    }

                    twsPlayerAttributes.gamesPlayedByPlayerCount[playerCount] += 1

                    int score = game.scores[(ObjectId) p.id]
                    if (score > twsPlayerAttributes.maxScoreByPlayerCount[playerCount]) {
                        twsPlayerAttributes.maxScoreByPlayerCount[playerCount] = score
                    }

                    if (game.winners.contains(gamePlayer.id)) {
                        twsPlayerAttributes.gamesWonByPlayerCount[playerCount] += 1
                    }

                    def saved = playerRepository.save(p)
                    playerPublisher.publish(saved)
            }
        }
    }
}
