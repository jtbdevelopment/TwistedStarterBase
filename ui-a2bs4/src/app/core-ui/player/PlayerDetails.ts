
export class GameTracker {
    freeGamesUsedToday: number;
    availablePurchasedGames: number;
    maxDailyFreeGames: number;
}

export class PlayerDetails {
    source: string;
    displayName: string;
    imageUrl: string;
    profileUrl: string;
    lastVersionNotes: string;
    adminUser: string;
    gameSpecificPlayerAttributes: GameTracker;
}