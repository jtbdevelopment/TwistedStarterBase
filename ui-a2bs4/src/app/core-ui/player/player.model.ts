
export class Player {
    id: string;
    md5: string;
    source: string;
    displayName: string;
    imageUrl: string;
    profileUrl: string;
    lastVersionNotes: string;
    adminUser: boolean = false;
    gameSpecificPlayerAttributes: any = {};

    constructor(copyFrom?: Player) {
        if (copyFrom) {
            this.id = copyFrom.id;
            this.md5 = copyFrom.md5;
            this.source = copyFrom.source;
            this.displayName = copyFrom.displayName;
            this.imageUrl = copyFrom.imageUrl;
            this.profileUrl = copyFrom.profileUrl;
            this.lastVersionNotes = copyFrom.lastVersionNotes;
            this.adminUser = copyFrom.adminUser;
            this.gameSpecificPlayerAttributes = Object.assign({}, copyFrom.gameSpecificPlayerAttributes);
        }
    }
}
