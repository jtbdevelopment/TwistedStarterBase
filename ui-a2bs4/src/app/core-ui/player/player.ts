
export class Player {
    source: string;
    displayName: string;
    imageUrl: string;
    profileUrl: string;
    lastVersionNotes: string;
    adminUser: boolean;
    //  TODO
    //gameSpecificPlayerAttributes: Map<string, string> = new Map<string, string>();
}
