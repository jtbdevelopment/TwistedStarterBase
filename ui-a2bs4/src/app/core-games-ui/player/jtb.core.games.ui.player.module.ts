import {NgModule} from '@angular/core';
import {HttpModule} from '@angular/http';
import {JTBCoreGamesUIMessageBus} from '../messagebus/jtb.core.games.ui.messagebus.module';
import {PlayerService} from './player.service';

@NgModule({
    imports: [HttpModule, JTBCoreGamesUIMessageBus],
    exports: [JTBCoreGamesUIMessageBus],
    providers: [
        PlayerService
    ]
})
export class JTBCoreGamesUIPlayer {
}
