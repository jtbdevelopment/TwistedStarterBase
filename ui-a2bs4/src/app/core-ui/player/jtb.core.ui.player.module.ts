import {NgModule} from '@angular/core';
import {HttpModule} from '@angular/http';
import {JTBCoreUIMessageBus} from '../messagebus/jtb.core.ui.messagebus.module';
import {PlayerService} from './player.service';

@NgModule({
    imports: [HttpModule, JTBCoreUIMessageBus],
    exports: [JTBCoreUIMessageBus],
    providers: [
        PlayerService
    ]
})
export class JTBCoreUIPlayer {
}
