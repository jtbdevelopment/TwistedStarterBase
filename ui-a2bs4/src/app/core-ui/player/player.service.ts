import {Injectable} from '@angular/core';
import {Player} from './player.model';
import {Observable, BehaviorSubject} from 'rxjs';
import {Http} from '@angular/http';
import {MessageBusService} from '../messagebus/message-bus.service';
import {Router} from '@angular/router';

@Injectable()
export class PlayerService {
    //  Unless dealing in admin functions use player
    //  In general, these two will be the same unless an admin player
    //  switches to simulation mode - then loggedInPlayer will retain the admin details
    //  while player will switch to the simulated player
    player: Observable<Player>;
    loggedInPlayer: Observable<Player>;

    private playerSubject: BehaviorSubject<Player> = new BehaviorSubject(new Player());
    private loggedInSubject: BehaviorSubject<Player> = new BehaviorSubject(new Player());

    //  TODO - not currently possible to test message bus updates when logged in != current
    constructor(private http: Http,
                private messageBus: MessageBusService,
                private router: Router) {
        this.player = Observable.from<Player>(this.playerSubject);
        this.loggedInPlayer = Observable.from<Player>(this.loggedInSubject);
        this.messageBus.playerUpdates.subscribe(player => {
            if (player.id === this.playerSubject.getValue().id) {
                console.log('player update ' + JSON.stringify(player));
                this.playerSubject.next(player);
            }
            if (player.id === this.loggedInSubject.getValue().id) {
                console.log('logged in update ' + JSON.stringify(player));
                this.loggedInSubject.next(player);
            }
        });
    }

    public loadLoggedInPlayer(): void {
        this.http.get('/api/security')
            .map(response => response.json())
            .map(json => {
                let loaded = new Player(json);
                return loaded;
            })
            .subscribe(loaded => {
                    this.playerSubject.next(loaded);
                    this.loggedInSubject.next(loaded);
                },
                error => {
                    //  TODO - general error handler
                    console.log(JSON.stringify(error));
                });
    }

    public logout(): void {
        this.http.post('/signout', null).subscribe(() => {
            this.resetPlayer();
        }, error => {
            console.log(JSON.stringify(error));
            this.resetPlayer();
        });
    }

    private resetPlayer(): void {
        this.loggedInSubject.next(new Player());
        this.playerSubject.next(new Player());
        this.messageBus.playerUpdates.next(new Player());
        this.router.navigateByUrl('/signin');
    }
}
