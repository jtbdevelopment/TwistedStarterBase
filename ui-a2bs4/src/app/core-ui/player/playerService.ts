import {Injectable} from '@angular/core';
import {PlayerDetails} from './PlayerDetails';
import {Observable, BehaviorSubject} from 'rxjs';

@Injectable()
export class PlayerService {
    //  Unless dealing in admin functions use player
    //  In general, these two will be the same, unless an admin player
    //  switches to simulation mode - then loggedInPlayer will retain the admin details
    //  while player will switch to the simulated player
    player: Observable<PlayerDetails>;
    loggedInPlayer: Observable<PlayerDetails>;

    private playerSubject: BehaviorSubject<PlayerDetails> = new BehaviorSubject(new PlayerDetails());
    private loggedInSubject: BehaviorSubject<PlayerDetails> = new BehaviorSubject(new PlayerDetails());

    constructor() {
        this.player = Observable.from<PlayerDetails>(this.playerSubject);
        this.loggedInPlayer = Observable.from<PlayerDetails>(this.loggedInSubject);
        //  TODO - eliminate
        let timer = Observable.timer(2000, 1000);
        timer.subscribe(t => {
            let p = new PlayerDetails();
            p.displayName = 'Testing';
            p.source = 'Manual';
            p.adminUser = true;
            this.playerSubject.next(p);
            this.loggedInSubject.next(p);
        });
    }
}
