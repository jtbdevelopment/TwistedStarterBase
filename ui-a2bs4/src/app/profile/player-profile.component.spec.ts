import {async, TestBed} from '@angular/core/testing';
import {PlayerService} from '../core-games-ui/player/player.service';
import {Player} from '../core-games-ui/player/player.model';
import {BehaviorSubject, Observable} from 'rxjs';
import {PlayerProfileComponent} from './player-profile.component';

export class MockPlayerService {
    static playerSubject: BehaviorSubject<Player> = new BehaviorSubject(new Player());

    player: Observable<Player> = Observable.from<Player>(MockPlayerService.playerSubject);
}

describe('player profile component', () => {
    let player = new Player({
        displayName: 'Name',
        source: 'test',
        imageUrl: '',
        gameSpecificPlayerAttributes: {freeGamesUsedToday: 20}
    });
    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [
                PlayerProfileComponent
            ],
            providers: [
                {provide: PlayerService, useClass: MockPlayerService}
            ],
        });
        TestBed.compileComponents();
        MockPlayerService.playerSubject.next(player);
    }));

    it('should render basics', () => {
        const fixture = TestBed.createComponent(PlayerProfileComponent);
        fixture.detectChanges();
        const navBar = fixture.nativeElement;
        expect(fixture.componentInstance.player).toEqual(player);
        let content = navBar.querySelector('table').textContent.trim();
        expect(content).toContain(player.displayName);
        expect(content).toContain(player.source);
        expect(content).toContain(player.gameSpecificPlayerAttributes.freeGamesUsedToday);
    });

    it('should update for a player', () => {
        const fixture = TestBed.createComponent(PlayerProfileComponent);
        fixture.detectChanges();
        const navBar = fixture.nativeElement;
        expect(fixture.componentInstance.player).toEqual(player);

        let newPlayer: Player = new Player({
            displayName: 'New Name',
            source: 'test2',
            imageUrl: '',
            gameSpecificPlayerAttributes: {freeGamesUsedToday: 30}
        });

        MockPlayerService.playerSubject.next(newPlayer);
        fixture.detectChanges();

        let content = navBar.querySelector('table').textContent.trim();
        expect(content).toContain(newPlayer.displayName);
        expect(content).toContain(newPlayer.source);
        expect(content).toContain(newPlayer.gameSpecificPlayerAttributes.freeGamesUsedToday);
    });
});
