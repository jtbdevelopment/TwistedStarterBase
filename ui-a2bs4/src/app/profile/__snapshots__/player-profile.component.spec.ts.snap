// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player profile component should render basics 1`] = `
<app-player-profile
  player={[Function Player]}
  playerService={[Function MockPlayerService]}
>
  <div
    class="player-profile"
  >
    
      
    <h2>
      Your profile:
    </h2>
    
      
    <table
      class="table table-hover table-sm text-right"
    >
      
          
      <tbody>
        <tr>
          
              
          <th>
            Name
          </th>
          
              
          <td>
            Name
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Logged In From
          </th>
          
              
          <td>
            test
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Type
          </th>
          
              
          <td>
            
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Daily Free Games
          </th>
          
              
          <td>
            
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Free Games Used Today
          </th>
          
              
          <td>
            20
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Free Games Remaining Today
          </th>
          
              
          <td>
            NaN
              
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Profile Image
          </th>
          
              
          <td>
            <img
              src=""
            />
          </td>
          
          
        </tr>
        
      
      </tbody>
    </table>
    
  
  </div>
  
</app-player-profile>
`;

exports[`player profile component should update for a player 1`] = `
<app-player-profile
  player={[Function Player]}
  playerService={[Function MockPlayerService]}
>
  <div
    class="player-profile"
  >
    
      
    <h2>
      Your profile:
    </h2>
    
      
    <table
      class="table table-hover table-sm text-right"
    >
      
          
      <tbody>
        <tr>
          
              
          <th>
            Name
          </th>
          
              
          <td>
            New Name
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Logged In From
          </th>
          
              
          <td>
            test2
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Type
          </th>
          
              
          <td>
            
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Daily Free Games
          </th>
          
              
          <td>
            
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Free Games Used Today
          </th>
          
              
          <td>
            30
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Free Games Remaining Today
          </th>
          
              
          <td>
            NaN
              
          </td>
          
          
        </tr>
        
          
        <tr>
          
              
          <th>
            Profile Image
          </th>
          
              
          <td>
            <img
              src=""
            />
          </td>
          
          
        </tr>
        
      
      </tbody>
    </table>
    
  
  </div>
  
</app-player-profile>
`;
